src/emulator/pubsubEmulator.ts

```diff
@@ -0,0 +1,186 @@
+    return javaEmulators.stop(Emulators.PUBSUB);
+  }
+
+  getInfo(): EmulatorInfo {
+    const host = this.args.host || Constants.getDefaultHost(Emulators.PUBSUB);
+    const port = this.args.port || Constants.getDefaultPort(Emulators.PUBSUB);
+
+    return {
+      host,
+      port,
+    };
+  }
+
+  getName(): Emulators {
+    return Emulators.PUBSUB;
+  }
+
+  async addTrigger(topicName: string, trigger: string) {
+    EmulatorLogger.logLabeled("DEBUG", "pubsub", `addTrigger(${topicName}, ${trigger})`);
+
+    const topicTriggers = this.triggers.get(topicName) || new Set();
+    if (topicTriggers.has(topicName) && this.subscriptions.has(topicName)) {
+      EmulatorLogger.logLabeled("DEBUG", "pubsub", "Trigger already exists");
+      return;
+    }
+
+    const topic = this.pubsub.topic(topicName);
+    try {
+      EmulatorLogger.logLabeled("DEBUG", "pubsub", `Creating topic: ${topicName}`);
+      await topic.create();
+    } catch (e) {
+      if (e && e.code === 6) {
+        EmulatorLogger.logLabeled("DEBUG", "pubsub", `Topic ${topicName} exists`);
+      } else {
+        throw new FirebaseError(`Could not create topic ${topicName}`, { original: e });
+      }
+    }
+
+    const subName = `emulator-sub-${topicName}`;
+    let sub;
+    try {
+      EmulatorLogger.logLabeled("DEBUG", "pubsub", `Creating sub for topic: ${topicName}`);
+      [sub] = await topic.createSubscription(subName);
+    } catch (e) {
+      if (e && e.code === 6) {
+        EmulatorLogger.logLabeled("DEBUG", "pubsub", `Sub for ${topicName} exists`);
+        sub = topic.subscription(`emulator-sub-${topicName}`);
+      } else {
+        throw new FirebaseError(`Could not create sub ${subName}`, { original: e });
+      }
+    }
+
+    sub.on("message", (message: Message) => {
+      this.onMessage(topicName, message);
+    });
+
+    topicTriggers.add(trigger);
+    this.triggers.set(topicName, topicTriggers);
+    this.subscriptions.set(topicName, sub);
+  }
+
+  private onMessage(topicName: string, message: Message) {
+    EmulatorLogger.logLabeled("DEBUG", "pubsub", `onMessage(${topicName}, ${message.id})`);
+    const topicTriggers = this.triggers.get(topicName);
+    if (!topicTriggers || topicTriggers.size === 0) {
+      throw new FirebaseError(`No trigger for topic: ${topicName}`);
+    }
+
+    let remaining = topicTriggers.size;
+    const postCallback = (err: any, res: request.Response) => {
+      if (err) {
+        EmulatorLogger.logLabeled(
+          "DEBUG",
+          "pubsub",
+          `Error running functions: ${JSON.stringify(err)}`
+        );
+      }
+
+      if (res) {
+        EmulatorLogger.logLabeled(
+          "DEBUG",
+          "pubsub",
+          `Functions emulator response: HTTP ${res.statusCode} ${JSON.stringify(res.body)}`
+        );
+      }
+
+      remaining--;
+      if (remaining <= 0) {
+        EmulatorLogger.logLabeled("DEBUG", "pubsub", `Acking message ${message.id}`);
+        message.ack();
+      }
```

Comment:
I don't understand what this code is doing. Can we add a comment explaining the workflow here?